WebCodec's VP8/VP9 VideoDecoder ignore video/full bitrange.

jyavenard opened the pull request:
#### 58383a97b1c609d4864328be6606b8ee946f6277
<pre>
WebCodec&apos;s VP8/VP9 VideoDecoder ignore video/full bitrange.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267400">https://bugs.webkit.org/show_bug.cgi?id=267400</a>
<a href="https://rdar.apple.com/120830372">rdar://120830372</a>

Reviewed by Youenn Fablet.

Pass the vpcC atom to the VideoDecoder and have LibWebRTCVPXVideoDecoder parse the VP&apos;s codec configuration.
to determine which CVPixelFormatType to use (FullRange or VideoRange).

If no vpcC is passed to the VideoDecoder at construction, we will retrieve the information from the VPx bytestream instead.

Added tests. While the tests do not exercise directly WebCodecs API per say, the VideoDecoder is used when
playing vp8 on iOS and we are indirectly testing the code.

* LayoutTests/media/content/test-h264-709-fullrange.mp4: Added.
* LayoutTests/media/content/test-h264-709-videorange.mp4: Added.
* LayoutTests/media/content/test-vp8-709-fullrange.webm: Added.
* LayoutTests/media/content/test-vp9-709-fullrange.webm: Added.
* LayoutTests/media/media-video-fullrange-expected.txt: Added.
* LayoutTests/media/media-video-fullrange.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations: WPE incorrectly renders all values as 0
* Source/WebCore/PlatformGTK.cmake: Add video_codes to search path.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/Headers.cmake: Added missing BitReader.h header.
* Source/WebCore/platform/graphics/VP9Utilities.cpp:
(WebCore::parseVPCodecParameters):
(WebCore::createVPCodecConfigurationRecordFromVPCC): Add method to parse VPCC box.
(WebCore::vpcCFromVPCodecConfigurationRecord): Extract code in its own method as it is needed to return information from the bytestream.
(WebCore::setConfigurationColorSpaceFromVP9ColorSpace):
(WebCore::vpcCFromVPXByteStream):
* Source/WebCore/platform/graphics/VP9Utilities.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
(WebCore::createVideoInfoFromVP9HeaderParser):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXInternalVideoDecoder::isVPx const):
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
(WebCore::LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder):
(WebCore::LibWebRTCVPXInternalVideoDecoder::createPixelBuffer):

Canonical link: <a href="https://commits.webkit.org/286305@main">https://commits.webkit.org/286305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/507d37908a4a9fd153ce1c747009dbdb00bd234e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75440 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79917 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77556 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2671 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59207 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17417 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/45 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39564 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22310 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25032 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81398 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2779 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1755 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67453 "Found 13 new test failures: imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-translate.html imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-001.html imported/w3c/web-platform-tests/css/css-view-transitions/far-away-capture.html imported/w3c/web-platform-tests/css/css-view-transitions/iframe-old-has-scrollbar.html imported/w3c/web-platform-tests/css/css-view-transitions/names-are-tree-scoped.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-captures-clip-path.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-object-view-box-clip-path-reference.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-object-view-box-clip-path.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-object-view-box-overflow.html imported/w3c/web-platform-tests/css/css-view-transitions/old-content-container-writing-modes.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2930 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66743 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16640 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10691 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8845 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2736 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5557 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2761 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3696 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2768 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
```diff
@@ -169,6 +170,7 @@ LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder(LibWebRTCVPXV
     : m_outputCallback(WTFMove(outputCallback))
     , m_internalDecoder(createInternalDecoder(type))
     , m_useIOSurface(config.pixelBuffer == VideoDecoder::HardwareBuffer::Yes)
+    , m_isFullRange((type == LibWebRTCVPXVideoDecoder::Type::VP8 || type == LibWebRTCVPXVideoDecoder::Type::VP9 || type == LibWebRTCVPXVideoDecoder::Type::VP9_P2) && config.description.size() >= 12 && (*(config.description.data() + 6) & 0x1))
```
This is rather cryptic, can we introduce a routine for checking full range and taking a span?
This will be useful when we move the VTB VP9 decoder in WebCore.
Also, I would use `config.description[6]` and maybe set it to a `bitDepthChromaAndRange` variable.

jernoble replied:
Additionally, this doesn't appear to handle the case that the color information is parsed out of the container file.

Can we just add `PlatformVideoColorSpace colorSpace;` to the `VideoDecoder::Config` struct?

jyavenard replied:
> Additionally, this doesn't appear to handle the case that the color information is parsed out of the container file.
> 
> Can we just add `PlatformVideoColorSpace colorSpace;` to the `VideoDecoder::Config` struct?

I'll take this as a follow-up, as it's a big change and require to modify the SourceBufferParserWebM to provide this information.

jyavenard replied:
I opened https://bugs.webkit.org/show_bug.cgi?id=282741 ; though I will be adding construction of the colorspace in this PR .

youennf reviewed (approved):
LGTM, but I would prefer that we fix `LibWebRTCVPXVideoDecoder` by getting the `vpcC` box from the stream itself.
This is what RTCVideoDecoderVTBVP9 does.

jyavenard committed:
WebCodec's VP8/VP9 VideoDecoder ignore video/full bitrange.
https://bugs.webkit.org/show_bug.cgi?id=267400
rdar://120830372

Reviewed by Youenn Fablet.

Pass the vpcC atom to the VideoDecoder and have LibWebRTCVPXVideoDecoder parse the VP's codec configuration.
to determine which CVPixelFormatType to use (FullRange or VideoRange).

If no vpcC is passed to the VideoDecoder at construction, we will retrieve the information from the VPx bytestream instead.

Added tests. While the tests do not exercise directly WebCodecs API per say, the VideoDecoder is used when
playing vp8 on iOS and we are indirectly testing the code.

* LayoutTests/media/content/test-h264-709-fullrange.mp4: Added.
* LayoutTests/media/content/test-h264-709-videorange.mp4: Added.
* LayoutTests/media/content/test-vp8-709-fullrange.webm: Added.
* LayoutTests/media/content/test-vp9-709-fullrange.webm: Added.
* LayoutTests/media/media-video-fullrange-expected.txt: Added.
* LayoutTests/media/media-video-fullrange.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations: WPE incorrectly renders all values as 0
* Source/WebCore/PlatformGTK.cmake: Add video_codes to search path.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/PlatformWin.cmake:
* Source/WebCore/Headers.cmake: Added missing BitReader.h header.
* Source/WebCore/platform/graphics/VP9Utilities.cpp:
(WebCore::parseVPCodecParameters):
(WebCore::createVPCodecConfigurationRecordFromVPCC): Add method to parse VPCC box.
(WebCore::vpcCFromVPCodecConfigurationRecord): Extract code in its own method as it is needed to return information from the bytestream.
(WebCore::setConfigurationColorSpaceFromVP9ColorSpace):
(WebCore::vpcCFromVPXByteStream):
* Source/WebCore/platform/graphics/VP9Utilities.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
(WebCore::createVideoInfoFromVP9HeaderParser):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXInternalVideoDecoder::isVPx const):
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
(WebCore::LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder):
(WebCore::LibWebRTCVPXInternalVideoDecoder::createPixelBuffer):

Canonical link: https://commits.webkit.org/286305@main
